### ISSUE TITLE
Multi Language Support

### ISSUE THREAD

**Issue by Teknoist** [lay295/TwitchDownloader#445]
### Checklist

- [X] I have checked the [issue page](https://github.com/lay295/TwitchDownloader/issues) for duplicates
- [X] I am running the latest version ([download here](https://github.com/lay295/TwitchDownloader/releases/latest))

### Write your feature request here

Can we add language support.

**Comment by ScrubN** [comment]
Looking into adding support for this. Translations would have to be community submitted though so it will take time

**Comment by ScrubN** [comment]
Good news, it should be possible to fully localize the CLI help text. I'll get around to adding localization after finishing the GUI frontend for the new chatupdate mode

**Comment by Teknoist** [comment]
I will be happy to help you translate into Turkish.

**Comment by ScrubN** [comment]
I'm taking a small break from programming while I wait for lay to review my open PRs, since 2 of them are pretty big changes. Once those have been merged, I'll continue working on translation support and I'll post here when it's ready for the actual translations.

When the time comes, it would be ideal for you guys submit PRs to my fork so the first few translations are bundled with my PR that adds translation support.

**Comment by ScrubN** [comment]
Apologies for the delay. It took me embarrassingly long to figure out I was missing the nuget package so I just put it off to the side. I'll probably have initial language support up in a few days if not today. I'll do my best to translate the app to french so you guys can see how to do it.

**Comment by ScrubN** [comment]
Initial support for the GUI is completed, though there are some outstanding GUI changes that will add more strings to be translated. I created an initial French translation as well specifically in its own commit (53245be1fe55768d525905e81cc2cd0c12469b03) so you guys can see the 3 steps needed to add a new language.

I recommend you use an editor like VSCode instead of full Visual Studio because VSCode does not auto generate designer.cs files and you can see when a key has a trailing space for the tooltip.

I will try to add translations to Core for the status/error messages but no guarantees that will work. I'm not really sure if translating the CLI will be super worth it to be honest. Translating the help text is a very different process than translating the GUI and Core and it's just a pain.

You can either wait for the PR to be merged to start working or you can submit a PR on my fork to the localization branch and it will be merged into the initial support PR.

**Comment by Teknoist** [comment]
There are still some English words in the French file, is it a necessity or a preference?
Did we have Strings.en.resx file little hard for me to France to Turkish.

![image](https://user-images.githubusercontent.com/37031361/212527068-fd54a616-4bdb-4c3f-9234-70237c68785c.png)

**Comment by ScrubN** [comment]
The English words are because I haven't done French since my first year of high school and am very rusty so I translated only the necessities. The only strings you can not translate have comments noting that they should not be translated.

[Strings.resx](https://github.com/ScrubN/TwitchDownloader/blob/localization-support/TwitchDownloaderWPF/Translations/Strings.resx) is the English translation, I probably should have clarified that because it is not included in the French commit.

**Comment by ScrubN** [comment]
Also if you want to make sure your translation looks correct in practice you will need to **publish** the application using Visual Studio or by following the [CLI build instructions](https://github.com/lay295/TwitchDownloader#building-from-source) because for whatever reason it does not work in debug/release builds.

**Comment by Teknoist** [comment]
![image](https://user-images.githubusercontent.com/37031361/212535975-06144f4a-df66-475b-aa50-209381ef5a75.png)
I translated all of them except the theme options.

**Comment by Teknoist** [comment]
Can you help me send the files? Github confuses me a bit.

**Comment by ScrubN** [comment]
Of course.

I would like you to create a new branch on your fork Teknoist/TwitchDownloader and commit everything you currently have locally to that branch.

Next you will need to go to my fork ScrubN/TwitchDownloader and click create a new pull request. From there you will see 2 dropdown menus with an arrow pointing from your fork to my fork. Select your branch with the changes and then select my branch named "localization support". From there I will be able merge your files into #508.

If you ever realize your need to make a change, you can simply make more commits to the same branch and it will automatically add those changes to your pull request. I will purposely wait to merge your pull request because of outstanding GUI changes that will add more text to be localized, which should let you practice committing after a pull request has been made.

Also I think you missed the update button ;)

**Comment by ScrubN** [comment]
Also the theme names cannot be localized because they pull from the file names of the themes and the literal word "System" is specifically checked for to determine automatic theming, making it possible to have the theme name be "System" in French but the theme service is checking for "System" in Turkish causing confusion and disabling auto-theming.

**Comment by Teknoist** [comment]
I tried to create a new branch, but there are no folders such as "AvailableCultures" file "Translations" in the branch

**Comment by ScrubN** [comment]
I assume what happened is you made the branch and your changes were automatically stashed. I don't use github desktop but I know Visual Studio 2022 has a stash explorer and I believe Visual Studio Code does too

**Comment by Teknoist** [comment]
https://github.com/lay295/TwitchDownloader/tree/master/TwitchDownloaderWPF
no folders such as  "Translations"

**Comment by ScrubN** [comment]
Also I just realized I should add a tooltip to the ffmpeg and file template parameters so there can be a localized explanation of what they do

**Comment by ScrubN** [comment]
> https://github.com/lay295/TwitchDownloader/tree/master/TwitchDownloaderWPF no folders such as "Translations"

Ah, you need to make the pull request on MY fork, not the master one. Don't worry, you will still be credited.

**Comment by Teknoist** [comment]
https://github.com/ScrubN/TwitchDownloader/pull/1

I'm sorry to bother you, I'm a little sick today, I realize some things little late :P

**Comment by ScrubN** [comment]
It's all good.

Like I said I will not merge your changes until the other GUI changes are merged so you may update your translation to include those

Congrats on figuring out the cross-fork pull request creator. It's a little unintuitive on your first go, I struggled with it too for a bit.

**Comment by Teknoist** [comment]
If you don't mind, I added a joke to the translation, instead of buy me coffe, I added buy me kebab.

**Comment by ScrubN** [comment]
That's fine by me, I see no harm in jokes as long as the original intent is still there

**Comment by ScrubN** [comment]
Just an update on the multi language support. It no longer requires a restart to apply the language and it also now works in debugging modes. Still waiting on the additional GUI changes and then I will expand the string list.

**Comment by Teknoist** [comment]
Can we use websites like Poeditor? 

For some people, it is difficult to translate with github, so I think they are afraid to translate it so that it will be a more comfortable translation.

https://poeditor.com/projects/

**Comment by ScrubN** [comment]
> Can we use websites like Poeditor?
> 
> For some people, it is difficult to translate with github, so I think they are afraid to translate it so that it will be a more comfortable translation.
> 
> https://poeditor.com/projects/

Not a bad idea but due to the extremely limited free plan we would absolutely need to apply for an open source license, which the prerequisites are not explained without making an account. 1,000 strings would only get us full translation in 3 languages - WPF is currently at 181 and if we support CLI & Core down the road that'll push us over 250 guaranteed.

Another option would be to use a different service like https://crowdin.com/ which would allow us to get probably 10 or more full languages out of the free plan, though its open source license terms would more than likely not accept this project if we needed to apply for it down the road.

**Comment by Teknoist** [comment]
https://tolgee.io/pricing/self-hosted self hosted version maybe help ?

**Comment by ScrubN** [comment]
> https://tolgee.io/pricing/self-hosted self hosted version maybe help ?

That application honestly looks really nice to work with. Unfortunately I can't currently afford to rent a public facing VPS for this, but if I did it would be hundreds of times cheaper than any of the paid plans the other services offer, assuming nobody tries to DDOS it. Maybe in the future we can use tolgee.

**Comment by ScrubN** [comment]
Initial localization support is now live! Thank you so much for the initial Turkish translation @Teknoist! CLI and Core (status/error message) translations are on the roadmap (see #575) but for now I would like to rewrite the task queue to address some user feedback & add CLI task queue support.
